sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java

```diff
@@ -196,11 +246,31 @@ Read withContext(ReaderContext context) {
     }
 
     @Override
+    @SuppressWarnings("deprecation")
     public PCollection<HCatRecord> expand(PBegin input) {
       checkArgument(getTable() != null, "withTable() is required");
       checkArgument(getConfigProperties() != null, "withConfigProperties() is required");
-
-      return input.apply(org.apache.beam.sdk.io.Read.from(new BoundedHCatalogSource(this)));
+      if (!getShouldTreatUnboundedAsBounded()) {
+        checkArgument(getPollingInterval() != null, "withPollingInterval() is required");
+        final Read readSpec =
+            new AutoValue_HCatalogIO_Read.Builder()
+                .setDatabase(DEFAULT_DATABASE)
+                .setShouldTreatUnboundedAsBounded(getShouldTreatUnboundedAsBounded())
+                .build();
+        return input
+            .apply("ConvertToReadRequest", Create.of(readSpec))
+            .apply(
+                "WatchForNewPartitions",
+                Watch.growthOf(new PartitionPollerFn(getConfigProperties()))
+                    .withPollInterval(getPollingInterval()))
+            .apply(Values.create())
+            .apply("PartitionSplitter", ParDo.of(new PartitionSplitterFn()))
+            .apply(Reshuffle.of())
+            .apply("PartitionReader", ParDo.of(new PartitionReaderFn()));
+      } else {
+        // Treat as Bounded
```

Comment:
If the only difference is that either we poll for new partitions or we don't, then we don't need shouldTreatUnboundedAsBounded, and should simply check whether pollingInterval is specified.